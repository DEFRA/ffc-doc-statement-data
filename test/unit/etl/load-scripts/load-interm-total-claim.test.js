const db = require('../../../../app/data')
const { loadIntermTotalClaim } = require('../../../../app/etl/load-scripts/load-interm-total-claim')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  }
}))

describe('loadIntermTotalClaim', () => {
  const startDate = '2023-01-01'
  const transaction = {}

  beforeEach(() => {
    db.sequelize.query.mockClear()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    await loadIntermTotalClaim(startDate, transaction)

    expect(db.sequelize.query).toHaveBeenCalledWith(`
    INSERT INTO etl_interm_total_claim (claim_id, payment_ref)
    SELECT
      (SELECT claim_id FROM etl_interm_finance_dax WHERE payment_ref = T.payment_ref LIMIT 1) as application_id,
      T.payment_ref
    FROM etl_interm_total T
    WHERE T.etl_inserted_dt > :startDate;
  `, {
      replacements: {
        startDate
      },
      raw: true,
      transaction
    })
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadIntermTotalClaim(startDate, transaction)).rejects.toThrow('Query failed')
  })
})
