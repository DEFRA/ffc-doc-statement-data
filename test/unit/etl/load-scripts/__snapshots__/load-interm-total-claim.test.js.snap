// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIntermTotalClaim should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    INSERT INTO public."etlIntermTotalClaim" ("claimId", "paymentRef")
    SELECT
      (SELECT "claimId" FROM public."etlIntermFinanceDax" WHERE "paymentRef" = T."paymentRef" LIMIT 1) as "applicationId",
      T."paymentRef"
    FROM public."etlIntermTotal" T
    WHERE T."etlInsertedDt" > :startDate
    ON CONFLICT ("claimId", "paymentRef") DO NOTHING;
  ",
      {
        "raw": true,
        "replacements": {
          "startDate": "2023-01-01",
        },
        "transaction": {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
